File: DFS&BFS/javascript/4803.js
const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");

const solution = () => {
  const answer = [];
  for (let i = 0; i < input.length - 2; ) {
    const [n, m] = input[i].split(" ").map(Number);
    const endIdx = i + m;
    const arr = Array.from(Array(n + 1), () => new Array());
    const visited = new Array(n + 1).fill(false);

    for (i++; i <= endIdx; i++) {
      const [a, b] = input[i].split(" ").map(Number);
      arr[a].push(b);
      arr[b].push(a);
    }

    let count = 0;
    for (let j = 1; j <= n; j++) {
      if (visited[j]) {
        continue;
      }
      if (dfs(-1, j, arr, visited)) {
        count++;
      }
    }
    answer.push(count);
  }
  return generateAnswer(answer);
};

const dfs = (prevNode, node, arr, visited) => {
  visited[node] = true;
  for (let nextNode of arr[node]) {
    if (nextNode === prevNode) {
      continue;
    }
    if (visited[nextNode]) {
      return false;
    }
    if (!dfs(node, nextNode, arr, visited)) {
      return false;
    }
  }
  return true;
};

const generateAnswer = (answer) => {
  return answer
    .map((cnt, idx) => {
      let template;
      if (cnt === 0) {
        template = "No trees.";
      } else {
        template = cnt > 1 ? `A forest of ${cnt} trees.` : "There is one tree.";
      }
      return `Case ${idx + 1}: ${template}`;
    })
    .join("\n");
};

const answer = solution();
console.log(answer);

File: DFS&BFS/python/4803.py
import sys
input = sys.stdin.readline
def dfs (prev_node,node,arr,visited):
  visited[node] = True
  for next_node in arr[node]:
    if (prev_node == next_node):
      continue
    if (visited[next_node]):
      return False
    if not dfs(node,next_node,arr,visited):
      return False
  return True
  

def printAnswer (answer):
  for i in range(1,len(answer)):
    cnt = answer[i]
    template = "No trees."
    if cnt>1:
      template = f"A forest of {cnt} trees."
    if cnt==1:
      template = "There is one tree."
    print(f"Case {i}: {template}")

answer = [None]
while True:
  N,M = map(int,input().split())
  if N == M == 0:
    break
  arr = [[] for _ in range(N+1)]
  visited = [False for _ in range(N+1)]
  cnt = 0
  for _ in range(M):
    a,b = map(int,input().split())
    arr[a].append(b)
    arr[b].append(a)

  for node in range(1,N+1):
    if (visited[node]):
      continue
    if dfs(-1,node,arr,visited):
      cnt += 1
  answer.append(cnt)

printAnswer(answer)